controllers/route/route_certificaterequest_controller.go

```diff
@@ -0,0 +1,362 @@
+			return false, time.Time{}
+		}
+		if !matches {
+			log.V(5).Info("Route private key doesn't match route cert")
+			return false, time.Time{}
+		}
+		// if we get here, we have a matching TLS cert/key pair.
+		// See if the cert will expire soon.
+		start := certs[0].NotBefore
+		end := certs[0].NotAfter
+		lifetime := end.Sub(start)
+		remaining := end.Sub(time.Now())
+		if remaining <= lifetime/3 {
+			return false, time.Time{}
+		}
+		return true, start.Add(2 * lifetime / 3)
+	}(instance)
+
+	if upToDate {
+		return c.ManageSuccessWithRequeue(ctx, instance, renewalTime.Sub(time.Now()))
+	}
+
+	// If we get here we need to fire off a certificate request. First step is to generate a new
+	// private key.
+	nextPrivateKeyPEM, found := instance.GetAnnotations()[cmapi.IsNextPrivateKeySecretLabelKey]
+	if !found {
+		// Generate a new private key, then re-reconcile.
+		// TODO: make key type and size configurable with an annotation
+		pk, err := cmutilpki.GenerateRSAPrivateKey(cmutilpki.MinRSAKeySize)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		pkPEM, err := cmutilpki.EncodePrivateKey(pk, cmapi.PKCS1)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		instance.GetAnnotations()[cmapi.IsNextPrivateKeySecretLabelKey] = string(pkPEM)
+		err = c.GetClient().Update(ctx, instance)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		return c.ManageSuccess(ctx, instance)
+	}
+	// Is the Private Key malformed?
+	pk, err := cmutilpki.DecodePrivateKeyBytes([]byte(nextPrivateKeyPEM))
+	if err != nil {
+		log.V(5).Error(err, "next private key annotation doesn't contain a valid key")
+		delete(instance.GetAnnotations(), cmapi.IsNextPrivateKeySecretLabelKey)
+		err = c.GetClient().Update(ctx, instance)
+		if err != nil {
+			return c.ManageError(ctx, instance, err)
+		}
+		return c.ManageSuccess(ctx, instance)
+	}
+
+	// We have a private key. Look for matching CertificateRequests
+	certificateRequests := &cmapi.CertificateRequestList{}
+	err = c.GetClient().List(ctx, certificateRequests, client.InNamespace(instance.Namespace))
+	if err != nil {
+		return c.ManageError(ctx, instance, err)
+	}
+	// only consider CertificateRequests that match the private key and are owned by this route
+	var ownedCRs []*cmapi.CertificateRequest
+	for _, cr := range certificateRequests.Items {
+		x509CR, err := x509.ParseCertificateRequest(cr.Spec.Request)
+		if err != nil {
+			continue
+		}
+		matches, err := cmutilpki.PublicKeyMatchesCSR(pk.Public(), x509CR)
+		if matches && err == nil {
+			for _, or := range cr.OwnerReferences {
+				if or.Kind == "Route" && or.Name == instance.Name {
+					ownedCRs = append(ownedCRs, cr.DeepCopy())
+					break
+				}
+			}
+		}
+	}
+	// only consider the most recent CR
+	sort.Slice(ownedCRs, func(i, j int) bool {
+		return ownedCRs[i].CreationTimestamp.Time.After(ownedCRs[j].CreationTimestamp.Time)
```

Comment:
Is the creation timestamp the best we can do? What about revision/not signed/signed with an expiry furthest in the future?